nameserver/dns.go

```diff
@@ -52,6 +62,9 @@ func NewDNSServer(ns *Nameserver, domain string, address string, ttl uint32, cli
 	if s.upstream, err = dns.ClientConfigFromFile(etcResolvConf); err != nil {
 		return nil, err
 	}
+	if s.upstream != nil {
```

Comment:
Maybe `if s.upstream != nil && len(effectiveAddress) > 0`, for saving some cycles...